File: db/golang/index-manager.go
package main

type IndexManager struct {
	index       map[string]int64
	fileAdapter *FileAdapter
}

func NewIndexManager(fa *FileAdapter) *IndexManager {
	return &IndexManager{fileAdapter: fa}
}

func (im *IndexManager) Init() error {
	index, err := im.fileAdapter.ReadIndex()
	if err != nil {
		return err
	}
	im.index = make(map[string]int64)
	for key, value := range index {
		if v, ok := value.(float64); ok {
			im.index[key] = int64(v)
		}
	}
	return nil
}

func (im *IndexManager) SetOffset(key string, offset int64) error {
	im.index[key] = offset
	return im.fileAdapter.SaveIndex(im.index)
}

func (im *IndexManager) GetOffset(key string) (int64, bool) {
	offset, exists := im.index[key]
	return offset, exists
}

File: db/nodejs/src/core/index-manager.ts
import { FileAdapter } from "./file.adapter";

export class IndexManager {
  private index: Map<string, { segmentNumber: number; offset: number }>

  constructor(protected fileAdapter: FileAdapter) {}

  async init() {
    this.index = await this.fileAdapter.readIndex();
  }

  async *readAll() {
      for (const [key, {offset, segmentNumber }] of this.index.entries()) {
        const data = await this.fileAdapter.get(offset, segmentNumber)
        yield { key, data: data };
      }
  }

  async setOffset(key: string, offset: number, segmentNumber: number) {
    this.index.set(key, {offset, segmentNumber});
    await this.fileAdapter.saveIndex(this.index);
  }

  async getOffset(key: string) {
    return this.index.get(key);
  }

  async delete(key: string) {
    this.index.delete(key);
    await this.fileAdapter.saveIndex(this.index);
  }
}
